src/StepperMotorReadback.cc

```diff
@@ -7,118 +7,175 @@
 
 #include "StepperMotorReadback.h"
 #include "ChimeraTK/MotorDriverCard/StepperMotorException.h"
+#include <mtca4u/MotorDriverCard/MotorDriverException.h>
 
 namespace ChimeraTK { namespace MotorDriver {
 
   void ReadbackHandler::mainLoop() {
     readConstData();
+    unsigned int spiErrorCounter = 0;
+    const unsigned int SPI_RETRY_COUNT = 50;
+
+    deviceError.status = static_cast<int32_t>(StatusOutput::Status::OK);
+    deviceError.message = "";
+    deviceError.setCurrentVersionNumber({});
+    deviceError.writeAll();
 
     while(true) {
       //First calculate the initial values
-
-      // FIXME This is only to evaluate the timer
-      if(!execTimer.isInitialized()) {
-        execTimer.initializeMeasurement();
+      try {
+        // FIXME This is only to evaluate the timer
+        if(!execTimer.isInitialized()) {
+          execTimer.initializeMeasurement();
+        }
+        else {
+          execTimer.measureIterativeMean();
+        }
+        auto ct = std::chrono::duration_cast<std::chrono::microseconds>(execTimer.getMeasurementResult());
+        actualCycleTime = static_cast<float>(ct.count()) / 1000.f;
+
+        receiveTimer.initializeMeasurement();
+
+        readback();
+        if(_motor->get()->hasHWReferenceSwitches()) {
+          readEndSwitchData();
+        }
+
+        receiveTimer.measureOnce();
+        auto rt = std::chrono::duration_cast<std::chrono::microseconds>(receiveTimer.getMeasurementResult());
+        actualReceiveTime = static_cast<float>(rt.count()) / 1000.f;
+
+        //We now have all data. Write them
+        writeAll();
+
+        // We have aparently communicated successfully. Reset the error counter
+        spiErrorCounter = 0;
       }
-      else {
-        execTimer.measureIterativeMean();
+      catch(mtca4u::MotorDriverException& e) {
+        if((e.getID() == mtca4u::MotorDriverException::SPI_ERROR ||
+               e.getID() == mtca4u::MotorDriverException::SPI_TIMEOUT)) {
+          if(spiErrorCounter > SPI_RETRY_COUNT) {
+            std::cerr << "SPI failed..." << std::endl;
+            incrementDataFaultCounter();
+            _motor->close();
+            deviceError.status = static_cast<int32_t>(StatusOutput::Status::FAULT);
+            deviceError.message = e.what();
+            deviceError.setCurrentVersionNumber({});
+            deviceError.writeAll();
+          }
+          else {
+            spiErrorCounter++;
+          }
+        }
+        else {
+          throw std::current_exception();
```

Comment:
This will re-throw a MotorDriverException that is not handled by ApplicationCore. So it basically means it's a logic error. Is this intended?

Reply:
Yes. The SPI errors are the only runtime errors, all other errors are non-recoverable. and configuration issues or calls of non-existent features

Reply:
Same as with the mtca4u include. the MotorDriverCard needs adaption. https://redmine.msktools.desy.de/issues/9678